Create a centralised cache manager

Chengxuan opened the pull request:
Create a centralised cache manager as described in #943

Chengxuan commented:
@awrichar comment addressed, will deal with the merge conflict next

Chengxuan commented:
@awrichar ready for review again.

Chengxuan reviewed (commented):
Ready to be reviewed again

peterbroadhurst commented:
I see the fabric tests failed - downloading looks to look for root cause

peterbroadhurst reviewed (changes requested):
Seeing this failure in the fabric tests on this line:

https://github.com/hyperledger/firefly/blob/e0aa9d78ee0e061efe9fc0a25e1cc3a12de39b57/internal/blockchain/fabric/fabric.go#L855

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0xb14a5d]

goroutine 8 [running]:
github.com/hyperledger/firefly/internal/blockchain/fabric.(*Fabric).GetNetworkVersion(0xc0004a4500, {0x103e9f0, 0xc000179780}, 0x96a5a5bc2e184b3e)
	/firefly/internal/blockchain/fabric/fabric.go:855 +0x9d
github.com/hyperledger/firefly/internal/multiparty.(*multipartyManager).configureContractCommon(0xc0004b1860, {0x103e9f0, 0xc000179780}, 0x0)
	/firefly/internal/multiparty/manager.go:152 +0x1fe
github.com/hyperledger/firefly/internal/multiparty.(*multipartyManager).ConfigureContract(0x1, {0x103e9f0, 0xc000179780})
	/firefly/internal/multiparty/manager.go:135 +0x25
github.com/hyperledger/firefly/internal/orchestrator.(*orchestrator).initManagers(0xc000220b40, {0x103e9f0, 0xc000179780})
	/firefly/internal/orchestrator/orchestrator.go:418 +0x317
github.com/hyperledger/firefly/internal/orchestrator.(*orchestrator).initComponents(0xc000220b40, {0x103e9f0, 0xc000179780})
	/firefly/internal/orchestrator/orchestrator.go:501 +0xe5
github.com/hyperledger/firefly/internal/orchestrator.(*orchestrator).Init(0xc000220b40, {0x103e9f0, 0xc000179740}, 0xc0004b5fa4)
	/firefly/internal/orchestrator/orchestrator.go:227 +0x12f
github.com/hyperledger/firefly/internal/namespace.(*namespaceManager).initNamespace(0xc0004c2370, {0x103e9f0, 0xc00047e000}, 0xc0004a4b00)
	/firefly/internal/namespace/manager.go:312 +0x5b4
github.com/hyperledger/firefly/internal/namespace.(*namespaceManager).Init(0xc0004c2370, {0x103e9f0, 0xc00047e000}, 0x1, 0xc000345f20)
	/firefly/internal/namespace/manager.go:223 +0x1e5
github.com/hyperledger/firefly/cmd.startFirefly({0x103e9f0, 0xc00047e000}, 0x0, {0x10526d0, 0xc0004c2370}, {0x1029140, 0xc00047f400}, 0x0, 0x0, 0xc000170060)
	/firefly/cmd/firefly.go:188 +0x49f
created by github.com/hyperledger/firefly/cmd.run
	/firefly/cmd/firefly.go:132 +0x4da
```

Chengxuan commented:
@peterbroadhurst thanks for the review. It’s caused by a mistake in the merge. I’ve fixed it and added more tests to catch it earlier. Spotted another error as well. Ready to be reviewed again

peterbroadhurst commented on internal/coremsgs/en_config_descriptions.go:
```diff
@@ -93,16 +93,32 @@ var (
+	ConfigCacheBatchTTL                = ffc("config.cache.batch.ttl", "Time to live of cache items for batches", i18n.StringType)
+	ConfigCacheBlockchainEventLimit    = ffc("config.cache.blockchainevent.limit", "Max number of cached blockchain events for transactions", i18n.IntType)
+	ConfigCacheBlockchainEventTTL      = ffc("config.cache.blockchainevent.ttl", "Time to live of cached blockchain events for transactions", i18n.StringType)
+	ConfigCacheTransactionSize         = ffc("config.cache.transaction.size", "Max size of cached transactions", i18n.IntType)
```
Want to just confirm this is a byte sized object (not a count).

peterbroadhurst commented on internal/coremsgs/en_config_descriptions.go:
```diff
@@ -93,16 +93,32 @@ var (
+	ConfigCacheIdentityTTL             = ffc("config.cache.identity.ttl", "Time to live of cached identities for identity manager", i18n.StringType)
+	ConfigCacheSigningKeyLimit         = ffc("config.cache.signingkey.limit", "Max number of cached signing keys for identity manager", i18n.IntType)
+	ConfigCacheSigningKeyTTL           = ffc("config.cache.signingkey.ttl", "Time to live of cached signing keys for identity manager", i18n.StringType)
+	ConfigCacheMessageSize             = ffc("config.cache.message.size", "Max size of cached messages for data manager", i18n.IntType)
```
See you have `i18n.IntType` here, but because it's a `.size` is will be `i18n.ByteSizeType`

Chengxuan commented on internal/cache/cache.go:
```diff
@@ -0,0 +1,222 @@
+	if err != nil {
+		return nil, err
+	}
+	cache, exists := cm.configuredCaches[cacheName]
```
this is a good point. I'll keep it in mind. Too used to node.js. Will add that in
